### PR TITLE
fix: remove expired saved cards

### PR DESCRIPTION
**Pull request by PritishBudhiraja**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

We were able to see saved cards which were expired. Filtered from fronted to don't see the expired cards.

Screenshot 2024-05-03 at 11 52 02 AM

## How did you test it?

Check for expired cards in locker and these shouldn't come in Saved PM Screen.

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by vsrivatsa25 — CHANGES_REQUESTED** [review]
Card is valid through the card.expiryMonth. Need to compare with next expiryMonth + 1 or expiryYear + 1 if expiryMonth == 12

File: src/PaymentElement.res [hunk 1]
```diff
@@ -61,7 +61,29 @@ let make = (~cardProps, ~expiryProps, ~cvcProps, ~paymentType: CardThemeType.mod
...
+              )->Int.toString
+
+            let expiryYear =
+              (
+                ele.card.expiryMonth->Int.fromString === Some(12)
+                  ? ele.card.expiryYear->Int.fromString->Option.getOr(0) + 1
+                  : ele.card.expiryYear->Int.fromString->Option.getOr(0)
+              )->Int.toString
+
+            let expiryDate = Date.fromString(`${expiryYear}-${expiryMonth}-01`)
```
Comment (vsrivatsa25) [inline src/PaymentElement.res:81]:
```suggestion
            let expiryDate = Date.fromString(`${expiryYear}-${expiryMonth}`)
```

File: src/PaymentElement.res [hunk 2]
```diff
@@ -61,7 +61,29 @@ let make = (~cardProps, ~expiryProps, ~cvcProps, ~paymentType: CardThemeType.mod
...
+            expiryDate > currentDate
```
Comment (vsrivatsa25) [inline src/PaymentElement.res:82]:
```suggestion
            expiryDate >= currentDate
```

**Review by vsrivatsa25 — APPROVED** [review]

File: src/Components/SavedCardItem.res [hunk 3]
```diff
@@ -49,6 +49,22 @@ let make = (
...
+    )->Int.toString
+  let expiryYear =
+    (
+      paymentItem.card.expiryMonth->Int.fromString === Some(12)
+        ? paymentItem.card.expiryYear->Int.fromString->Option.getOr(0) + 1
+        : paymentItem.card.expiryYear->Int.fromString->Option.getOr(0)
+    )->Int.toString
+  let expiryDate = Date.fromString(`${expiryYear}-${expiryMonth}`)
+  let currentDate = Date.make()
+  let isCardExpired = isCard ? expiryDate < currentDate : false
```
Comment (vsrivatsa-juspay) [inline src/Components/SavedCardItem.res:66]:
```suggestion
  let isCardExpired = isCard && expiryDate < currentDate
```

File: src/Components/SavedCardItem.res [hunk 4]
```diff
@@ -123,37 +142,44 @@ let make = (
...
+                    maxLength=4
+                    inputRef=cvcRef
+                    placeholder="123"
+                    height="2.2rem"
+                  />
                 </div>
-              </RenderIf>
+              </div>
+            </RenderIf>
+            <RenderIf condition={isCard && isCardExpired}>
```
Comment (vsrivatsa-juspay) [inline src/Components/SavedCardItem.res:175]:
```suggestion
            <RenderIf condition={isCardExpired}>
```
isCardExpired already checks off isCard

**Review by vsrivatsa-juspay — APPROVED** [review]

**Review by ArushKapoorJuspay — APPROVED** [review]

**Review by ArushKapoorJuspay — APPROVED** [review]

**Pull request merged**